cpp/ql/src/Likely Bugs/Underspecified Functions/ImplicitFunctionDeclaration.qlhelp

```diff
@@ -0,0 +1,28 @@
+<!DOCTYPE qhelp PUBLIC
+  "-//Semmle//qhelp//EN"
+  "qhelp.dtd">
+<qhelp>
+
+
+<overview>
+<p>A function is called without a prior function declaration or definition.
+When this happens, the compiler generates an implicit declaration of the function,
+specifying an integer return type and no parameters.</p>
+</p>
+
+<p>This may indicate a misspelled function name, or that the required header containing
+the function declaration has not been included.</p>
```

Comment:
The consequences of these alerts sound benign in the qhelp, so I'm tempted to downgrade the severity to `recommendation` if we can't find any worse consequences.

In my reading of https://en.cppreference.com/w/c/language/type#Compatible_types and C11 6.5.2.2, it's undefined behaviour to call an implicitly-declared function if any of its parameters are affected by default argument promotions. For example, the implementation of a `void myfunc(float)` would be allowed to find its argument in a 32-bit stack location, but calling `myfunc(1.0f)` without the prototype in scope would promote the argument to a `double` and therefore must place it in a location that's at least 64 bits wide.

@dbartol, what's your interpretation of the C standard here? If there are potential issues on real compilers, that would of course be even better than theoretical issues with the standard.

Reply:
The example from https://wiki.sei.cmu.edu/confluence/display/c/DCL31-C.+Declare+identifiers+before+using+them shows a really bad case of this:

```
#include <stddef.h>
/* #include <stdlib.h> is missing */
  
int main(void) {
  for (size_t i = 0; i < 100; ++i) {
    /* int malloc() assumed */
    char *ptr = (char *)malloc(0x10000000);
    *ptr = 'a';
  }
  return 0;
}
```

Here, only half of the pointer returned by `malloc` would be copied to `ptr` (assuming a 32 bit integer and 64 bit pointer), which could make `ptr` point to invalid memory causing undefined behavior for `*ptr = 'a';`

Reply:
Nice. Although I guess that only works because the C programmer in this case has followed their C++ habits and has written a `(char *)` cast on the `malloc`. Otherwise the code wouldn't compile. But that could still happen in real code.

Reply:
Good point. I guess you can craft something almost as bad without requiring an explicit conversion like this:

```
// File: 1.c:
double returns_double() {
  return 3.3;
}
```

```
// File: 2.c:
#include <stdio.h>

// double returns_double();

int main() {
  int r;
  r = returns_double();
  printf("%d\n", r);
  return 0;
}
```
If the programmer forgets to write a declaration for `returns_double` the compiler won't generate an implicit conversion from double to int, and `r` will contain garbage:

```
> gcc -std=c89 1.c 2.c && ./a.out
48914176
```

Reply:
Okay, so the risks are real. To justify the 'warning' severity, you'll need to allude to these risks in the qhelp. I personally like the `malloc` example best because it's contained in one file.